Sources/Logging/MetadataProvider.swift

```diff
@@ -0,0 +1,123 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Logging API open source project
+//
+// Copyright (c) 2018-2022 Apple Inc. and the Swift Logging API project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift Logging API project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
+import Darwin
+#elseif os(Windows)
+import CRT
+#elseif canImport(Glibc)
+import Glibc
+#elseif canImport(WASILibc)
+import WASILibc
+#else
+#error("Unsupported runtime")
+#endif
+
+#if compiler(>=5.6)
+@preconcurrency protocol _SwiftLogSendable: Sendable {}
+#else
+protocol _SwiftLogSendable {}
+#endif
+
+extension Logger {
+    /// A `MetadataProvider` is used to automatically inject runtime-generated metadata
+    /// to all logs emitted by a logger.
+    ///
+    /// ### Example
+    /// A metadata provider may be used to automatically inject metadata such as
+    /// trace IDs:
+    ///
+    /// ```swift
+    /// import Tracing // https://github.com/apple/swift-distributed-tracing
+    ///
+    /// let metadataProvider = MetadataProvider {
+    ///     guard let traceID = Baggage.current?.traceID else { return nil }
+    ///     return ["traceID": "\(traceID)"]
+    /// }
+    /// let logger = Logger(label: "example", metadataProvider: metadataProvider)
+    /// var baggage = Baggage.topLevel
+    /// baggage.traceID = 42
+    /// Baggage.withValue(baggage) {
+    ///     logger.info("hello") // automatically includes ["traceID": "42"] metadata
+    /// }
+    /// ```
+    ///
+    /// We recommend referring to [swift-distributed-tracing](https://github.com/apple/swift-distributed-tracing)
+    /// for metadata providers which make use of its tracing and metadata propagation infrastructure. It is however
+    /// possible to make use of metadata providers independently of tracing and instruments provided by that library,
+    /// if necessary.
+    public struct MetadataProvider: _SwiftLogSendable {
+        /// A default no-op metadata provider, which always returns empty metadata.
+        public static var noop: MetadataProvider {
```

Comment:
can remove this now that provider is optional?

Reply:
We can I guess, it'll force writing the same `MetadataProvider { [:] }` by us or users sometimes... if we frantically want to avoid a "noop" API then sure we can remove this...